File: app/src/main/java/me/liam/support/SupportTransaction.java
package me.liam.support;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.animation.AnimationUtils;

import java.util.List;
import java.util.UUID;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import me.liam.anim.FragmentAnimation;
import me.liam.helper.FragmentUtils;
import me.liam.queue.Action;
import me.liam.queue.ActionQueue;

public class SupportTransaction {

    final public static String FRAGMENTATION_CONTAINER_ID = "Fragmentation:ContainerId";
    final public static String FRAGMENTATION_TAG = "Fragmentation:Tag";
    final public static String FRAGMENTATION_SIMPLE_NAME = "Fragmentation:SimpleName";
    final public static String FRAGMENTATION_FULL_NAME = "Fragmentation:FullName";
    final public static String FRAGMENTATION_PLAY_ENTER_ANIM = "Fragmentation:PlayEnterAnim";
    final public static String FRAGMENTATION_INIT_LIST = "Fragmentation:InitList";
    final public static String FRAGMENTATION_BACK_STACK = "Fragmentation:AddToBackStack";

    final public static String FRAGMENTATION_ENTER_ANIM_ID = "Fragmentation:EnterAnimId";
    final public static String FRAGMENTATION_EXIT_ANIM_ID = "Fragmentation:ExitAnimId";
    final public static String FRAGMENTATION_POP_ENTER_ANIM_ID = "Fragmentation:PopEnterAnimId";
    final public static String FRAGMENTATION_POP_EXIT_ANIM_ID = "Fragmentation:PopExitAnimId";

    final public static String FRAGMENTATION_SAVED_INSTANCE = "Fragmentation:SavedInstance";

    final public static String FRAGMENTATION_REQUEST_CODE = "Fragmentation:RequestCode";
    final public static String FRAGMENTATION_FROM_REQUEST_CODE = "Fragmentation:FromRequestCode";
    final public static String FRAGMENTATION_RESULT_CODE = "Fragmentation:ResultCode";
    final public static String FRAGMENTATION_RESULT_DATA = "Fragmentation:ResultData";

    private ISupportActivity iSupportActivity;

    private FragmentActivity supportActivity;

    private ActionQueue actionQueue;

//    private Context context;

    SupportTransaction(ISupportActivity iSupportActivity) {
        this.iSupportActivity = iSupportActivity;
        this.supportActivity = (FragmentActivity) iSupportActivity;
        actionQueue = new ActionQueue(new Handler(Looper.myLooper()));

    }

    public Bundle getArguments(SupportFragment target){
        if (target.getArguments() == null){
            target.setArguments(new Bundle());
        }
        return target.getArguments();
    }

    private void bindFragmentOptions(SupportFragment target, int containerId, boolean playEnterAnim, boolean addToBackStack){
        Bundle args = getArguments(target);
        args.putInt(FRAGMENTATION_CONTAINER_ID, containerId);
        args.putString(FRAGMENTATION_TAG, UUID.randomUUID().toString());
        args.putString(FRAGMENTATION_SIMPLE_NAME, target.getClass().getSimpleName());
        args.putString(FRAGMENTATION_FULL_NAME, target.getClass().getName());
        args.putBoolean(FRAGMENTATION_PLAY_ENTER_ANIM, playEnterAnim);
        args.putBoolean(FRAGMENTATION_INIT_LIST, false);
        args.putBoolean(FRAGMENTATION_SAVED_INSTANCE,false);
        args.putBoolean(FRAGMENTATION_BACK_STACK,addToBackStack);
    }

    private void supportCommit(FragmentTransaction ft) {
        supportCommit(ft,null);
    }

    private void supportCommit(FragmentTransaction ft,Runnable runnable) {
        if (runnable != null){
            ft.runOnCommit(runnable);
        }
        ft.commitAllowingStateLoss();
    }

    void loadRootFragment(final FragmentManager fm, final int containerId, final SupportFragment to, final FragmentAnimation anim, final boolean playEnterAnim, final boolean addToBackStack){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                bindFragmentOptions(to,containerId,playEnterAnim,addToBackStack);
                to.setFragmentAnimation(anim);
                FragmentTransaction ft = fm.beginTransaction();
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                ft.add(containerId,to);
                supportCommit(ft);
                return 0;
            }
        });
    }

    void loadMultipleRootFragments(final FragmentManager fm, final int containerId, final int showPosition, final SupportFragment... fragments){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                int position = 1;
                FragmentTransaction ft = fm.beginTransaction();
                for (SupportFragment to : fragments){
                    bindFragmentOptions(to,containerId,false,false);
                    to.setFragmentAnimation(null);
                    ft.add(containerId,to);
                    if (position == showPosition){
                        ft.show(to);
                    }else {
                        ft.hide(to);
                    }
                }
                supportCommit(ft);
                return 0;
            }
        });
    }

    void showHideAllFragment(final FragmentManager fm, final SupportFragment show){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = fm.beginTransaction();
                for (SupportFragment f : FragmentUtils.getInManagerFragments(fm)){
                    if (f == show){
                        ft.show(f);
                    }else {
                        ft.hide(f);
                    }
                }
                supportCommit(ft);
                return 0;
            }
        });
    }

    void start(final SupportFragment from, final SupportFragment to,final boolean addToBackStack){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                bindFragmentOptions(to,from.getContainerId(),true,addToBackStack);
                to.setFragmentAnimation(iSupportActivity.getDefaultAnimation());
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                ft.add(to.getContainerId(),to);
                supportCommit(ft);
                return 0;
            }
        });
    }


    void pop(final FragmentManager fm){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                SupportFragment remove = FragmentUtils.getLastFragment(fm);
                onResult(remove);
                if (remove == null) return 0;
                long duration = AnimationUtils.loadAnimation(remove.getContext(),remove.getFragmentAnimation().getExitAnimId()).getDuration();
                FragmentTransaction ft = fm.beginTransaction();
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                ft.remove(remove);
                supportCommit(ft);
                return duration;
            }

            @Override
            public int actionType() {
                return Action.TYPE_POP;
            }
        });
    }

    void startWithPop(final SupportFragment from, final SupportFragment to){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                bindFragmentOptions(to,from.getContainerId(),true,true);
                to.setFragmentAnimation(iSupportActivity.getDefaultAnimation());
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                ft.add(to.getContainerId(),to);
                supportCommit(ft);
                to.setCallBack(new SupportFragmentCallBack(){
                    @Override
                    public void onEnterAnimEnd() {
                        silenceRemove(from.getFragmentManager(),from);
                    }
                });
                return 0;
            }
        });
    }

    void silenceRemove(final FragmentManager fm, final SupportFragment... removes){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = fm.beginTransaction();
                for (SupportFragment f : removes){
                    ft.remove(f);
                }
                supportCommit(ft);
                return 0;
            }
        });
    }

    void popTo(final FragmentManager fm, final Class cls, final boolean includeTarget){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                SupportFragment remove = FragmentUtils.getLastFragment(fm);
                SupportFragment target = FragmentUtils.findFragmentByClass(fm,cls);
                if (remove == null || target == null) return 0;
                FragmentTransaction ft = fm.beginTransaction();
                int targetIndex = fm.getFragments().indexOf(target);
                int removeIndex = fm.getFragments().indexOf(remove);
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                List<Fragment> removeList = fm.getFragments().subList(targetIndex,removeIndex);
                if (!includeTarget){
                    removeList.remove(target);
                }
                for (Fragment f : removeList){
                    if (f instanceof SupportFragment){
                        ft.remove(f);
                    }
                }
                ft.remove(remove);
                supportCommit(ft);
                return 0;
            }

            @Override
            public int actionType() {
                return Action.TYPE_POP;
            }
        });
    }

    void startForResult(SupportFragment from, SupportFragment to, int requestCode){
        Bundle formBundle = getArguments(from);
        formBundle.putInt(FRAGMENTATION_REQUEST_CODE,requestCode);
        Bundle toBundle = getArguments(to);
        toBundle.putInt(FRAGMENTATION_FROM_REQUEST_CODE,requestCode);
        start(from, to,true);
    }

    void setResult(SupportFragment target, int resultCode, Bundle data){
        Bundle bundle = getArguments(target);
        bundle.putInt(FRAGMENTATION_RESULT_CODE,resultCode);
        bundle.putBundle(FRAGMENTATION_RESULT_DATA,data);
    }

    void onResult(SupportFragment target){
        if (target == null) return;
        Bundle bundle = getArguments(target);
        int fromRequestCode = bundle.getInt(FRAGMENTATION_FROM_REQUEST_CODE,-1);
        int resultCode = bundle.getInt(FRAGMENTATION_RESULT_CODE,-1);
        if (resultCode == -1 || fromRequestCode == -1) return;
        Bundle data = bundle.getBundle(FRAGMENTATION_RESULT_DATA);
        if (target.getFragmentManager() == null) return;
        List<SupportFragment> list = FragmentUtils.getActiveList(target.getFragmentManager());
        for (SupportFragment f : list){
            if (getArguments(f).getInt(FRAGMENTATION_REQUEST_CODE) == fromRequestCode){
                f.onResult(fromRequestCode,resultCode,data);
            }
        }
    }

    void popAll(final FragmentManager fm){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = fm.beginTransaction();
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                for (Fragment f : fm.getFragments()){
                    if (f instanceof SupportFragment && !f.isRemoving() && !f.isDetached()){
                        ft.remove(f);
                    }
                }
                supportCommit(ft);
                return 0;
            }

            @Override
            public int actionType() {
                return Action.TYPE_POP;
            }
        });
    }

    void remove(final SupportFragment remove, final boolean anim){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                if (remove == null || remove.getFragmentManager() == null) return 0;
                FragmentTransaction ft = remove.getFragmentManager().beginTransaction();
                if (anim){
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                }
                ft.remove(remove);
                supportCommit(ft);
                return 0;
            }

            @Override
            public int actionType() {
                return Action.TYPE_POP;
            }
        });
    }

    void startWithPopTo(final SupportFragment from, final SupportFragment to,final Class cls, final boolean includeTarget){
        actionQueue.enqueue(new Action() {
            @Override
            public long run() {
                FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                bindFragmentOptions(to,from.getContainerId(),true,true);
                to.setFragmentAnimation(iSupportActivity.getDefaultAnimation());
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                ft.add(to.getContainerId(),to);
                supportCommit(ft);
                to.setCallBack(new SupportFragmentCallBack(){
                    @Override
                    public void onEnterAnimEnd() {
                        popTo(from.getFragmentManager(),cls,includeTarget);
                    }
                });
                return 0;
            }

            @Override
            public int actionType() {
                return Action.TYPE_POP;
            }
        });
    }
}

File: core/src/main/java/me/liam/fragmentation/support/SupportTransaction.kt
package me.liam.fragmentation.support

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.animation.AnimationUtils
import java.util.UUID

import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import me.liam.fragmentation.anim.FragmentAnimation
import me.liam.fragmentation.helper.*
import me.liam.fragmentation.queue.Action
import me.liam.fragmentation.queue.ActionQueue

class SupportTransaction internal constructor(private val iSupportActivity: ISupportActivity) {

    private val actionQueue = ActionQueue(Handler(Looper.getMainLooper()))

    private fun getArguments(target: SupportFragment): Bundle {
        return target.arguments ?: Bundle().also { target.arguments = it }
    }

    private fun bindFragmentOptions(target: SupportFragment, containerId: Int, playEnterAnim: Boolean, addToBackStack: Boolean) {
        val args = getArguments(target)
        args.putInt(FRAGMENTATION_CONTAINER_ID, containerId)
        args.putString(FRAGMENTATION_TAG, UUID.randomUUID().toString())
        args.putString(FRAGMENTATION_SIMPLE_NAME, target.javaClass.simpleName)
        args.putString(FRAGMENTATION_FULL_NAME, target.javaClass.name)
        args.putBoolean(FRAGMENTATION_PLAY_ENTER_ANIM, playEnterAnim)
        args.putBoolean(FRAGMENTATION_INIT_LIST, false)
        args.putBoolean(FRAGMENTATION_SAVED_INSTANCE, false)
        args.putBoolean(FRAGMENTATION_BACK_STACK, addToBackStack)
    }

    internal fun loadRootFragment(fm: FragmentManager, containerId: Int, to: SupportFragment, anim: FragmentAnimation, playEnterAnim: Boolean, addToBackStack: Boolean) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                bindFragmentOptions(to, containerId, playEnterAnim, addToBackStack)
                to.fragmentAnimation = anim

                fm.beginTransaction().apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    add(containerId, to)
                }.supportCommit()
                return 0
            }
        })
    }

    internal fun loadMultipleRootFragments(fm: FragmentManager, containerId: Int, showPosition: Int, vararg fragments: SupportFragment) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                var position = 1
                val ft = fm.beginTransaction()
                for (to in fragments) {
                    bindFragmentOptions(to, containerId, false, false)
                    to.fragmentAnimation = null

                    ft.add(containerId, to)
                    if (position == showPosition) {
                        ft.show(to)
                    } else {
                        ft.hide(to)
                    }
                    position++
                }
                ft.supportCommit()
                return 0
            }
        })
    }

    internal fun showHideAllFragment(fm: FragmentManager, show: SupportFragment) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                val ft = fm.beginTransaction()
                for (f in fm.getInManagerFragments()) {
                    if (f === show) {
                        ft.show(f)
                    } else {
                        ft.hide(f)
                    }
                }
                ft.supportCommit()
                return 0
            }
        })
    }

    internal fun start(from: SupportFragment?, to: SupportFragment, addToBackStack: Boolean) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                from ?: return 0

                bindFragmentOptions(to, from.containerId, true, addToBackStack)
                to.fragmentAnimation = iSupportActivity.defaultAnimation

                from.fragmentManager?.beginTransaction()?.apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    add(to.containerId, to)
                }?.supportCommit()
                return 0
            }
        })
    }


    internal fun pop(fm: FragmentManager) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                val remove = fm.getLastFragment() ?: return 0

                val duration = remove.fragmentAnimation?.exitAnimId?.let { AnimationUtils.loadAnimation(remove.context, it).duration }
                fm.beginTransaction().apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                    remove(remove)
                }.supportCommit()
                return duration ?: 0
            }

            override fun actionType(): Int {
                return Action.TYPE_POP
            }
        })
    }

    internal fun startWithPop(from: SupportFragment?, to: SupportFragment) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                from ?: return 0

                bindFragmentOptions(to, from.containerId, true, true)
                to.fragmentAnimation = iSupportActivity.defaultAnimation

                from.fragmentManager?.beginTransaction()?.apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    add(to.containerId, to)
                }?.supportCommit()

                to.setAnimEndListener {
                    silenceRemove(from.fragmentManager, from)
                }
                return 0
            }
        })
    }

    internal fun silenceRemove(fm: FragmentManager?, vararg removes: SupportFragment) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                fm?.beginTransaction()?.apply {
                    for (f in removes) remove(f)
                }?.supportCommit()
                return 0
            }
        })
    }

    internal fun popTo(fm: FragmentManager, cls: Class<*>, includeTarget: Boolean) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                val remove = fm.getLastFragment()
                val target: SupportFragment? = fm.findFragment(cls)
                if (remove == null || target == null) return 0

                val targetIndex = fm.fragments.indexOf(target)
                val removeIndex = fm.fragments.indexOf(remove)

                val removeList = fm.fragments.subList(targetIndex, removeIndex)
                val ft = fm.beginTransaction()
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                if (!includeTarget) {
                    removeList.remove(target)
                }
                removeList.filter { it is SupportFragment }.forEach { ft.remove(it) }
                ft.remove(remove)
                ft.supportCommit()
                return 0
            }

            override fun actionType(): Int {
                return Action.TYPE_POP
            }
        })
    }

    internal fun startForResult(from: SupportFragment, to: SupportFragment, requestCode: Int) {
        getArguments(from).apply { putInt(FRAGMENTATION_REQUEST_CODE, requestCode) }
        getArguments(to).apply { putInt(FRAGMENTATION_FROM_REQUEST_CODE, requestCode) }
        start(from, to, true)
    }

    internal fun setResult(target: SupportFragment, resultCode: Int, data: Bundle) {
        val bundle = getArguments(target)
        bundle.putInt(FRAGMENTATION_RESULT_CODE, resultCode)
        bundle.putBundle(FRAGMENTATION_RESULT_DATA, data)
    }

    internal fun onResult(target: SupportFragment?) {
        target ?: return

        val bundle = getArguments(target)
        val fromRequestCode = bundle.getInt(FRAGMENTATION_FROM_REQUEST_CODE, -1)
        val resultCode = bundle.getInt(FRAGMENTATION_RESULT_CODE, -1)
        if (resultCode == -1 || fromRequestCode == -1) return

        val data = bundle.getBundle(FRAGMENTATION_RESULT_DATA)

        val list = target.fragmentManager?.getActiveList()
        list?.filter {
            getArguments(it).getInt(FRAGMENTATION_REQUEST_CODE) == fromRequestCode
        }?.forEach { it.onResult(fromRequestCode, resultCode, data) }
    }

    internal fun popAll(fm: FragmentManager) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                fm.beginTransaction().apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                    fm.fragments.filter {
                        it is SupportFragment && !it.isRemoving() && !it.isDetached()
                    }.forEach { remove(it) }
                }.supportCommit()
                return 0
            }

            override fun actionType(): Int {
                return Action.TYPE_POP
            }
        })
    }

    internal fun remove(remove: SupportFragment?, anim: Boolean) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                remove?.fragmentManager?.beginTransaction()?.apply {
                    if (anim) setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                    remove(remove)
                }?.supportCommit()
                return 0
            }

            override fun actionType(): Int {
                return Action.TYPE_POP
            }
        })
    }

    internal fun startWithPopTo(from: SupportFragment?, to: SupportFragment, cls: Class<*>, includeTarget: Boolean) {
        actionQueue.enqueue(object : Action() {
            override fun run(): Long {
                from ?: return 0

                bindFragmentOptions(to, from.containerId, true, true)
                to.fragmentAnimation = iSupportActivity.defaultAnimation

                from.fragmentManager?.beginTransaction()?.apply {
                    setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    add(to.containerId, to)
                }?.supportCommit()

                to.setAnimEndListener {
                    from.fragmentManager?.let { popTo(it, cls, includeTarget) }
                }
                return 0
            }

            override fun actionType(): Int {
                return Action.TYPE_POP
            }
        })
    }

    companion object {

        const val FRAGMENTATION_CONTAINER_ID = "Fragmentation:ContainerId"
        const val FRAGMENTATION_TAG = "Fragmentation:Tag"
        const val FRAGMENTATION_SIMPLE_NAME = "Fragmentation:SimpleName"
        const val FRAGMENTATION_FULL_NAME = "Fragmentation:FullName"
        const val FRAGMENTATION_PLAY_ENTER_ANIM = "Fragmentation:PlayEnterAnim"
        const val FRAGMENTATION_INIT_LIST = "Fragmentation:InitList"
        const val FRAGMENTATION_BACK_STACK = "Fragmentation:AddToBackStack"

        const val FRAGMENTATION_ENTER_ANIM_ID = "Fragmentation:EnterAnimId"
        const val FRAGMENTATION_EXIT_ANIM_ID = "Fragmentation:ExitAnimId"
        const val FRAGMENTATION_POP_ENTER_ANIM_ID = "Fragmentation:PopEnterAnimId"
        const val FRAGMENTATION_POP_EXIT_ANIM_ID = "Fragmentation:PopExitAnimId"

        const val FRAGMENTATION_DONT_DISPLAY_SELF_POP_ANIM = "Fragmentation:DontDisplaySelfPopAnim"

        const val FRAGMENTATION_SAVED_INSTANCE = "Fragmentation:SavedInstance"

        const val FRAGMENTATION_REQUEST_CODE = "Fragmentation:RequestCode"
        const val FRAGMENTATION_FROM_REQUEST_CODE = "Fragmentation:FromRequestCode"
        const val FRAGMENTATION_RESULT_CODE = "Fragmentation:ResultCode"
        const val FRAGMENTATION_RESULT_DATA = "Fragmentation:ResultData"
    }
}
